SDCARDDETECTINVERTED problems

eboston opened the pull request:
Fixed an error where SDCARDDETECTINVERTED was being defined/undefined
improperly

eboston committed:
SDCARDDETECTINVERTED problems

Fixed an error where SDCARDDETECTINVERTED was being defined/undefined
improperly

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -289,7 +289,9 @@
-  #define SDCARDDETECTINVERTED
+  #ifndef ELB_FULL_GRAPHIC_CONTROLLER
+    #define SDCARDDETECTINVERTED
+  #endif
```
Instead of this being here, it should be in `Conditionals.h` as the opposite case:

```
#ifdef ELB_FULL_GRAPHIC_CONTROLLER
  #undef SDCARDDETECTINVERTED
#endif
```

thinkyhead commented:
Your idea may be sound, but I wanted to figure out how that value might end up being set in the first place.

Here's a mapping of how `SDCARDDETECTINVERTED` gets set and unset…

In the first include of `Conditionals.h` there is:

```
  #if defined(miniVIKI) || defined(VIKI2) || defined(ELB_FULL_GRAPHIC_CONTROLLER)
    #ifdef miniVIKI
    #elif defined(VIKI2)
    #elif defined(ELB_FULL_GRAPHIC_CONTROLLER)
      #define SDCARDDETECTINVERTED
```

But, if `SDSUPPORT` is enabled, `SDCARDDETECTINVERTED` is enabled by `Configuration_adv.h` no matter:

```
#ifdef SDSUPPORT
  #define SDCARDDETECTINVERTED
  ...
```

But then, as you point out in those lines, `SDCARDDETECTINVERTED` gets disabled again (after `Configuration_adv.h` gets processed) if `ULTIPANEL` is enabled.

By the way, `ULTIPANEL` is enabled under these conditions:

```
#if defined(MAKRPANEL)
#if defined(ULTIMAKERCONTROLLER) || defined(REPRAP_DISCOUNT_SMART_CONTROLLER) || defined(G3D_PANEL)
#ifdef REPRAPWORLD_KEYPAD
#ifdef RA_CONTROL_PANEL
#ifdef LCD_I2C_SAINSMART_YWROBOT
#ifdef LCD_I2C_PANELOLU2
#ifdef LCD_I2C_VIKI
#ifdef SAV_3DLCD
```

Note that when `ELB_FULL_GRAPHIC_CONTROLLER` is enabled, it causes `ULTIMAKERCONTROLLER` to be enabled, and this in turn causes `ULTIPANEL` to be enabled. The change to `Conditionals.h` in this PR prevents `SDCARDDETECTINVERTED` being disabled when `ELB_FULL_GRAPHIC_CONTROLLER` is enabled.

One thing that confuses me is that this PR also prevents `SDCARDDETECTINVERTED` being enabled in the first place, in `Configuration_adv.h`.

It would be nice to just clean up this twisty maze of display flags, but we may be stuck with it for a while.

eboston commented:
I have been fighting with these twisty maze defines this whole project.  In regards to your first message, yes, it should be in Conditionals.h.  It is something that should be defined by the controller.  In my case, I needed it defined.  I put the change in Configuration_adv.h so it would not effect any other display.  Since I don't have all the other displays and platforms, I can't test out the impact of just removing it.

In regards to the second message.  SDCARDDETECTINVERTED should be defined by the controller.  What is happening in these other areas don't match up with what you described to me with the first PR I did.  Just because SDSUPPORT is defined doesn't mean SDCARDDETECTINVERTED should be.  That should be determined by the controller implementation and the type of detection used by the card socket.  In my case, the full size SD card socket I used on my PCB design has a different detection than the microSD socket I use on my breadboard.  When I started working on the next changes, I commented out the define under ELB_FULL_GRAPHIC_CONTROLLER, yet it didn't change anything.  I added these conditionals to remove the assumptions being made by the code so it would only be changed if using my controller.

From my experience with the code, all the defines in regards to the LCD controllers need to be reworked.

So what do you want me to do in regards to this PR?

thinkyhead commented:
> What do you want me to do in regards to this PR?

@eboston I think all you need is the original change to `Conditionals.h` to prevent `SDCARDDETECTINVERTED` from being disabled when `ELB_FULL_GRAPHIC_CONTROLLER` is in use. One of these days one of us will do a complete map of all the options and how they relate to each other. It wasn't too tricky to trace the happenings in this situation, just have to do the same thing x50.

eboston commented:
Not having the change in Configuration_adv.h would break the ability to turn on and off the feature as needed in my debugging.  If I disabled it in Conditionals.h, it would get turned right back on in Configuration_adv.h and then my breadboard system would not work.

thinkyhead commented:
The _second half_ of `Conditionals.h` is interpreted _after_ `Configuration_adv.h`. So any code you place there can do what needs to be done.

eboston commented:
But you can't really tell that by looking at the one file.  And if things get shuffled, it could break.  It wouldn't be a complete fix without it and from what it sounds like you are saying, it really doesn't hurt to have it in.

thinkyhead commented:
Gosh, there has to be some clean way to do it without adding this conditional about a specific board into `Configuration_adv.h`. We've been so good about moving everything else like it to `Conditionals.h` :eyes:.

> ability to turn on and off the feature as needed in my debugging

I see your point. If it was a "true" or "false" value instead of an "ifdef / ifndef" value then it wouldn't be so much trouble.

thinkyhead commented:
Here's another option, based on that thought: Define it as "0" (off) or "1" (on) from your Makefile, and we can use `#if` instead of `#ifdef` for testing it. Undefined symbols are treated as "0" by the preprocessor.

```
#if !UNDEFINED_THING
  #warning "This works!"
#endif
```

Then in `Configuration_adv.h` wrap it like:

```
#ifndef SDCARDDETECTINVERTED
  #define SDCARDDETECTINVERTED 1
#endif
```

eboston commented:
The clean way would be to remove the define completely from Conditionals_adv.h and only have it in the device section like I tried to do.  Both these places I put the changes in don't seem to fit your moving everything to Conditionals.h desire.  It could be done in "theory", but to be sure, all configurations would need to be tested.

Using the #if, while it would work, would make it different from the others and I thought the idea was to make it all uniform.  Unless I'm not understanding what you are trying to say.

thinkyhead commented:
@eboston I'm trying too hard to be clever. All these approaches have their own special pitfalls. There is still a cleaner way, once my brain fog clears...

eboston commented:
The clean way of doing it is to remove the define in `Conditionals_adv.h`.  That would remove the need for my change.

thinkyhead commented:
Ok, really we should be doing this with all our switches:

```
#if !defined(SDCARDDETECTINVERTED) || !SDCARDDETECTINVERTED
  ...
#endif
```

And we should add a macro for that purpose:

```
#define ENABLED(OPT) (!(!defined(OPT) || !OPT))
```

Which makes any such option overridable from `Makefile`:

```
#if !ENABLED(SDCARDDETECTINVERTED)
  ...
#endif
```

...assuming the compiler properly ignores `#define MY_SETTING` on the second go-round as it should – but we can wrap them all in `#ifndef` all over the place, one supposes… (It's always been a can of worms to make things easily overridable.)

thinkyhead commented:
> The clean way of doing it is to remove the define in `Conditionals_adv.h`.

So, move it to `Conditionals.h` and hide it from configuration altogether?

eboston commented:
Doing either of these leaves us with a testing issue.  While I can test that it works on my system, it may break other controllers.  If you are willing to take that responsibility, check in a change that moves it so my change won't be needed.

Doing the macros for new defines would make sense.  The next set of changes I am working on does introduce some new defines so I could do those this new way.  But doing existing one carries a risk of breaking existing setups.

thinkyhead commented:
> it may break other controllers

It would almost be okay within the context of Marlin and the configs that come with it, just doing the equivalent thing in `Conditionals.h`:

```
#ifdef SDSUPPORT
  #define SDCARDDETECTINVERTED
#endif
```

Except, this would not be compatible with the Makibox, which has it commented out in its `Configuration_adv.h`.

eboston commented:
Even moving it there, the fact that `SDCARDDETECTINVERTED` is hard coded to be defined if `SDSUPPORT` is, is the real problem.  Even if it was moved, I would want my change it just so it was clear that it is not defined by default for my controller.

thinkyhead commented:
A pattern: When `SDSLOW` is enabled by a configuration `SDCARDDETECTINVERTED` is being disabled too. Coincidence?

thinkyhead commented:
> `SDCARDDETECTINVERTED` is hard coded to be defined if `SDSUPPORT` is

Not hard-coded, precisely. Its inclusion in `Configuration_adv.h` is an invitation to change it to suit your hardware.

thinkyhead commented:
No guarantee that this is always true, but it also prevents the option from being disabled for that controller:

```
#if defined(ULTIPANEL) && !defined(SDSLOW)
  #undef SDCARDDETECTINVERTED
#endif
```

eboston commented:
So you want me to change what does work to something that should work?  While it may work now, it might now in the future.  One thing I am working on is fixing the need to use SDSLOW.  So once I get that resolved, we are right back to where we were.  That fix wouldn't work and we would need to come up with another band-aide to apply.

I understand your desire for having the separation, and I agree fully with it.  But until all the code base is fixed to support that, these types of fixes are going to be needed.

thinkyhead commented:
> you want me to change what does work to something that should work?

Just floating concepts here. Maybe we can gain more insight into the reasoning behind making `SDCARDDETECTINVERTED` a primary option. I presume it's so tinkerers can try their custom controllers with Marlin, while the known displays perhaps should just set it as part of their conditionals.

eboston commented:
Floating concepts is fine, but I feel it should be done in a different thread and not hold up my changes in the meantime.

As for SDCARDDETECTINVERTED, being a primary option, it would depend on what your definition of a primary option is.  It is a definition that describes the configuration of the controller.  It could be argued that it is a subset of SDSUPPORT, but it should still only be defined under the controller's entry in Conditionals.h.  The code should not be making assumptions like it is.  But until this is fixed in the code, these types of conditions are going to be required for new devices.

thinkyhead commented:
> not hold up my changes in the meantime

Sorry, I'm doing my best to rationalize it, but I don't think this PR is going to be merged as-is. We may need input from more developers to move it forward.

You may need to change this line too: https://github.com/eboston/Marlin/blob/Development/Marlin/Conditionals.h#L36 It always defines `SDCARDDETECTINVERTED` for `ELB_FULL_GRAPHIC_CONTROLLER` no matter what you do in `Configuration_adv.h`.

eboston commented:
Huh?  It does not always define it for ELB_FULL_GRAPHIC_CONTROLLER.  That is a "if not defined" conditional.  It only gets set if it ISN'T my controller, which is what the code currently does.  It is because of that define in Configuration_adv.h that my setting in Conditionals.h doesn't work.  It overrides any setting I make.  I just ran a test to verify that Configuration_adv.h is being processed after my settings in Configurations.h.  The blind setting of SDCARDDECTECTINVERTED in Configuration_adv.h is a bug and is causing the setting that have been made they way they are suppose to be, not work.  My fix is limited to only my controller so it doesn't break anything that is currently out there.  That is why it is where it is and how it is.  I wish I didn't have to do it, but that is the only way I can see to get my code to work without the possibility of breaking other configurations.

eboston commented:
That last message got messed up.  The line in Conditionals.h you are referring to is how it is suppose to be done.  For my controller, by default, it should always be set that way.  However, for my prototyping, I need to turn it off, but because of the two places in the code I made changes to, it overrides any changes I make.

NarimaanV commented:
Gonna jump in here to share my experiences with this as I might be the only other person who's been playing around with this display and making a controller for it. I've been working with @eboston (and by "working with" I mean he's been helping me get my controller together and working) for a couple months about this and we saw recently for certain card readers, the detection inversion messed up card detection. Good example was my card reader I'm using for my controller needs the detection to NOT be inverted (like @eboston 's prototyping card reader), so we went into conditionals.h and commented it out for the display and saw that that did absolutely nothing, leading to @eboston to make these changes which fixed card detection for card readers like mine. Now I don't know which case is most prevalent of which controllers need detection inversion, if the majority do so it should be enabled by default, or if they don't do it shouldn't, but there has to be someplace the user should be able to enable or disable that feature, without having to make changes to multiple files, which at the moment isn't true. The current layout apparently doesn't work otherwise @eboston wouldn't need this PR. So my two cents, not being a collab or even an expert coder, allow this PR at least temporarily until a better solution can be found. Not a ton of people use this display for this purpose so it won't even affect many users (only know of two so far). But like I said, that's just my two cents.

thinkyhead commented:
> we went into conditionals.h and commented it out for the display and saw that that did absolutely nothing

That's right. Because it was subsequently being set in `Configuration_adv.h`. You should have also commented out the line in `Configuration_adv.h`, preventing it from being defined there.

thinkyhead commented:
> there has to be someplace the user should be able to enable or disable that feature

That is correct. It should only _need_ to be set in `Configuration_adv.h`. The problem is that it is being defined in `Conditionals.h` when `ELB_FULL_GRAPHIC_CONTROLLER` is used. That line in `Conditionals.h` should probably be removed altogether. It's getting in the way of being able to set it in `Configuration_adv.h`.

thinkyhead commented:
> I don't know which case is most prevalent of which controllers need detection inversion

Me neither, but the longstanding default for `SDCARDDETECTINVERTED` has been: enabled when `SDSUPPORT` is enabled. But later, as this PR points out, there is a longstanding piece of code (moved from `Configuration_adv.h` over to `Conditionals.h` during re-organization)…

```
  #ifdef ULTIPANEL
    #undef SDCARDDETECTINVERTED
  #endif
```

Which disables the option for some reason, always, when `ULTIPANEL` is enabled. As I traced out, `ULTIPANEL` is enabled for `ELB_FULL_GRAPHIC_CONTROLLER`, so no matter what you do ahead of this line, `SDCARDDETECTINVERTED` will be turned off. If (as it appears) this should not be done with the `ELB_FULL_GRAPHIC_CONTROLLER` then there are some options.
- Add the condition as outlined in the PR so we get…

```
  #if defined(ULTIPANEL) && !defined(ELB_FULL_GRAPHIC_CONTROLLER)
    #undef SDCARDDETECTINVERTED
  #endif
```
- or, Put the original `#undef` conditional before the line that defines it for `ELB_FULL_GRAPHIC_CONTROLLER` (or move the forced-enable to a later point), because (apparently) it is _intended_ that `SDCARDDETECTINVERTED` should always be set with `ELB_FULL_GRAPHIC_CONTROLLER`.

eboston commented:
Which of the options will get the PR merged?

thinkyhead commented:
> Huh? It does not always define it for ELB_FULL_GRAPHIC_CONTROLLER

At the top of `Conditionals.h` follow the cascade, and you can see, a line has been added to ensure that `SDCARDDETECTINVERTED` will always be enabled when `ELB_FULL_GRAPHIC_CONTROLLER` is enabled (unless miniVIKI or VIKI2):

```
  #if defined(miniVIKI) || defined(VIKI2) || defined(ELB_FULL_GRAPHIC_CONTROLLER)
    #ifdef miniVIKI
    #elif defined(VIKI2)
    #elif defined(ELB_FULL_GRAPHIC_CONTROLLER)
      #define SDCARDDETECTINVERTED
```

But it doesn't work, because later (after `Configuration_adv.h` has been processed) it's disabled again.

```
  #ifdef ULTIPANEL
    #undef SDCARDDETECTINVERTED
  #endif
```

thinkyhead commented:
> Which of the options will get the PR merged?

@eboston If your intent is to make sure that `SDCARDDETECTED` is always enabled for the `ELB_FULL_GRAPHIC_CONTROLLER` then it should be explicitly set in the second half of `Conditionals.h` so that nothing else can override it. For example:

```
  #ifdef ELB_FULL_GRAPHIC_CONTROLLER
    #define SDCARDDETECTINVERTED
  #elif defined(ULTIPANEL)
    #undef SDCARDDETECTINVERTED
  #endif
```

If your intent is simply to allow users to set `SDCARDDETECTED` as usual according to their preferences, just remove the line that forces it to be set for `ELB_FULL_GRAPHIC_CONTROLLER`.

```
    #elif defined(ELB_FULL_GRAPHIC_CONTROLLER)
      #define SDCARDDETECTINVERTED // remove me!
```

If your intent is to make sure that `SDCARDDETECTED` is never enabled for the `ELB_FULL_GRAPHIC_CONTROLLER` then you don't need to do anything (except remove the line that forces it to be set for `ELB_FULL_GRAPHIC_CONTROLLER`). It will be unset because `ULTIPANEL`.

eboston commented:
The intent is to allow the user to define or not define it as needed.  The default is to have it defined as that is the method needed for the PCB that I have developed.  But it needs to be able to be undefined (comment it out) for microSD card sockets used in the breadboard.

thinkyhead commented:
> The default is to have it defined … for the PCB that I have developed

Great. Currently `SDCARDDETECTINVERTED` is enabled with `SDSUPPORT` unless this is changed. The use of an `ULTIPANEL` controller forces `SDCARDDETECTINVERTED` to be disabled.

For the ELB controller we need an exception to this behavior. We need `SDCARDDETECTINVERTED` to remain user-configurable, and for that value to stick, and not to be disabled just because `ELB_FULL_GRAPHIC_CONTROLLER` causes `ULTIPANEL` to be defined.

Simple enough.

We should remove the line that forces it to always be enabled for `ELB_FULL_GRAPHIC_CONTROLLER`:

```
  #elif defined(ELB_FULL_GRAPHIC_CONTROLLER)
    #define SDCARDDETECTINVERTED // remove me!
```

Then we should add the additional condition preventing it from being unset by `ULTIPANEL`:

```
#if defined(ULTIPANEL) && !defined(ELB_FULL_GRAPHIC_CONTROLLER)
  #undef SDCARDDETECTINVERTED
#endif
```

That's all. We can leave `Configuration_adv.h` as it is.

eboston commented:
And what does the user do when they need it not defined?

thinkyhead commented:
> And what does the user do when they need it not defined?

If it is commented out in `Configuration_adv.h` it will not be subsequently defined anywhere.

eboston commented:
So instead of being able to define/undefine an option under the definitions of the controller, the user is suppose to know to go hunt down a define they don't know about (not listed with the other options for the controller) and comment it out?

thinkyhead commented:
> instead of being able to define/undefine an option under the definitions of the controller

Controllers don't have configurable options currently. `SDCARDDETECTINVERTED` is an SD-related option, not a controller-specific option.

thinkyhead commented:
If your controller has special needs, you can add configuration files to `example_configurations` to demonstrate the settings it requires. There is some precedent for that approach. See `example_configurations/RepRapWorld/Megatronics/Configuration.h` for example.

eboston commented:
It is a SD-related option, but one that needs to be set based on the hardware the controller is using, not because SD functionality is enabled.

thinkyhead commented:
@eboston I agree that the organization needs to be better. But this is how we do it for the codebase we're stuck with at the moment. I rolled #2154 as the example for you.

The `SDCARDDETECTINVERTED` setting has (up to now) been exposed as something users may need to change if their SD card detect line doesn't seem to work. If it can be given a known proper value based on the controller in use, that's great. Unfortunately so far no one has done the work to make that explicit in the code (except for making it on by default, and forced off for `ULTIPANEL`, of course).

eboston commented:
Maybe I have been looking at this wrong.  It sounds to me you are saying the user should never change anything in `Conditionals.h`.  Is that right?  You want that to stay as static as possible.  If there are tweaks that need to be made, it needs to be done in Configuration.h.  So maybe a solution is to add a special define in Configuration.h like REPRAPWORLD_KEYPAD_MOVE_STEP 10.0.  I could add a define like ELB_FULL_GRAPHIC_CONTROLLER_MICROSD that indicates the controller is using a microSD card and the detection doesn't need to be inverted.  

And with the next changes I have coming, I can add another define like ELB_FULL_GRAPHIC_cONTROLLER_OPTION_1_INSTALLED and use that for where the changes need to be made.  But this options is something other LCD displays may want to use so shouldn't be be defined differently?

thinkyhead commented:
> the user should never change anything in `Conditionals.h`

Correct. Sorry, I know there should be better documentation in `Conditionals.h`. The idea has been to make it so you only need to modify `Configuration.h` and `Configuration_adv.h` to configure any machine. The `Conditionals.h` and `SanityCheck.h` files were added to offload the parts of those files that were never meant to change and to provide standard places for those invariant parts.

> So maybe a solution is to add a special define in Configuration.h like `REPRAPWORLD_KEYPAD_MOVE_STEP 10.0`. I could add a define like `ELB_FULL_GRAPHIC_CONTROLLER_MICROSD` that indicates the controller is using a microSD card and the detection doesn't need to be inverted.

That's one possible solution. But then, if your controller doesn't need `SDCARDDETECTINVERTED` to be enabled (or rather, if it requires it to be _disabled_), then that is already the case, because `ULTIPANEL` forces it to be disabled. Are there some cases where you want the user to be able to set it?

---

The `SDCARDDETECTINVERTED` setting has never been placed very well for cases where it needs to be _enabled with `ULTIPANEL`_, and the current arrangement is really just like the old one. We can possibly do something to improve the placement of this setting. The old code only did this: https://github.com/MarlinFirmware/Marlin/blob/Release/Marlin/Configuration_adv.h#L376

The comments there indicate that `SDCARDDETECTINVERTED` should be left commented out in most cases, but of course the default was to have it enabled for all controllers except those that set `ULTIPANEL`. So, effectively it is _disabled_ in most cases. And the assumption seems to be that users will rarely need to change it.

eboston commented:
>  Are there some cases where you want the user to be able to set it?

Yes.  As I have been saying, depending on the SD card socket used, detect may, or may not, need to be enabled.  Since this is a display that others could build, the type of socket used cannot be predicted.  I wanted to set the default to the way my PCB is going to implement it.  But someone else may not use it.

thinkyhead commented:
Your option to override seems like a good idea. Instead of `ELB_FULL_GRAPHIC_CONTROLLER_MICROSD` maybe something shorter. `ELB_USING_MICROSD` is easy to understand and can be placed right after the switch. Or, if we just want to be straightforward about what the option does, you could name it `ELB_SDCARDDETECTINVERTED` and it can just always override `SDCARDDETECTINVERTED`. In that case the block in Part 2 of `Conditionals.h` would become:

```
#ifdef ULTIPANEL
  #ifdef ELB_FULL_GRAPHIC_CONTROLLER
    #ifdef ELB_SDCARDDETECTINVERTED
      #define SDCARDDETECTINVERTED
    #else
      #undef SDCARDDETECTINVERTED
    #endif
  #else
    #undef SDCARDDETECTINVERTED
  #endif
#endif
```

…or for short…

```
#ifdef ULTIPANEL
  #undef SDCARDDETECTINVERTED
  #if defined(ELB_FULL_GRAPHIC_CONTROLLER) && defined(ELB_SDCARDDETECTINVERTED)
    #define SDCARDDETECTINVERTED
  #endif
#endif
```

eboston commented:
I was thinking of shortening the actual name to.  Change it to

ELB_GLCD

And for the detect

ELB_GLCD_NO_SDCARDDETECTINVERTED

Or something like that.  I haven't had much time to look or even think about it lately.  Thinking maybe to just close this PR and do another one when things are ready.

thinkyhead commented:
> ELB_GLCD

If `GLCD` was a standard mnemonic for "Graphical LCD" it would make more sense. But `ELB_FULL_GRAPHIC_CONTROLLER` is a good name, consistent with the other long names for controllers, which makes it stand out as it should.

> ELB_GLCD_NO_SDCARDDETECTINVERTED

I am against having options that disable things. It goes against longstanding guidelines in the UX world about how options should be implemented. Instead of having an option that _disables_ something and having that enabled by default, it is always better to have an option that _enables_ something and have it disabled by default. This avoid confusion around double-negatives, and just makes the verbiage clearer. For this reason, using a `ELB_SDCARDDETECTINVERTED` setting would be a more favored way to approach it.

eboston committed:
Added ELB_FULL_GRAPHIC_CONTROLLER_SDCARDDETECTINVERTED

The new define allows the sd card detect to be inverted or not from
Configuration.h

eboston commented:
Okay.  Let's see if this captures what you are wanting.  The new define allows the detect to be inverted or not from Configuration.h.  The original changes still need to be in place so they don't overwrite any changes I make in regards to the display type.

EDIT:  Oh, and I know the change needs to be added to the other Configuration.h files.  I wanted to find out if this is correct before making those changes.

eboston commented:
Any comment on this?

eboston committed:
Removed SDSLOW for ELB_FULL_GRAPHIC_CONTROLLER

Hardware changes has removed the need for this option.

eboston committed:
Adding ELB_FULL_GRAPHIC_CONTROLLER_SDCARDDETECTINVERTED to Examples
